test/topology/test_cluster_features.py

```diff
@@ -0,0 +1,163 @@
+    """Simulates an upgrade of a cluster by doing a rolling restart
+       and marking the test-only feature as supported on restarted nodes.
+    """
+    servers = await manager.running_servers()
+
+    cql = manager.cql
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+
+    for srv in servers:
+        host = (await wait_for_cql_and_get_hosts(cql, [srv], time.time() + 60))[0]
+
+        # The feature should not be advertised as supported until we enable it
+        # via error injection.
+        assert TEST_FEATURE_NAME not in await get_supported_features(cql, host)
+
+        # Until all nodes are updated to support the test feature, that feature
+        # should not be considered enabled by any node.
+        for host in hosts:
+            assert TEST_FEATURE_NAME not in await get_enabled_features(cql, host)
+
+        await turn_on_test_feature_and_restart(manager, srv)
+        cql = await reconnect_driver(manager)
+
+        host = (await wait_for_cql_and_get_hosts(cql, [srv], time.time() + 60))[0]
+        assert TEST_FEATURE_NAME in await get_supported_features(cql, host)
+
+    # The feature should become enabled on all nodes soon.
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+    logging.info(f"Waiting until {TEST_FEATURE_NAME} is enabled on all nodes")
+    await asyncio.gather(*(wait_for_feature(TEST_FEATURE_NAME, cql, h, time.time() + 60) for h in hosts))
+
+
+@pytest.mark.asyncio
+async def test_partial_cluster_upgrade_then_downgrade(manager: ManagerClient) -> None:
+    """Simulates an partial of a cluster by enabling the test features in all
+       but one nodes, then downgrading them.
+    """
+    servers = await manager.running_servers()
+    upgrading_servers = servers[1:] if len(servers) > 0 else []
+
+    cql = manager.cql
+
+    # Upgrade
+    for srv in upgrading_servers:
+        await turn_on_test_feature_and_restart(manager, srv)
+        cql = await reconnect_driver(manager)
+        host = (await wait_for_cql_and_get_hosts(cql, [srv], time.time() + 60))[0]
+        assert TEST_FEATURE_NAME in await get_supported_features(cql, host)
+
+    # There is one node that is not upgraded. The feature should not be enabled.
+    for srv in servers:
+        cql = await reconnect_driver(manager)
+        host = (await wait_for_cql_and_get_hosts(cql, [srv], time.time() + 60))[0]
+        assert TEST_FEATURE_NAME not in await get_enabled_features(cql, host)
+
+    # Downgrade, in reverse order
+    for srv in upgrading_servers[::-1]:
+        await turn_off_test_feature_and_restart(manager, srv)
+        cql = await reconnect_driver(manager)
+        host = (await wait_for_cql_and_get_hosts(cql, [srv], time.time() + 60))[0]
+        assert TEST_FEATURE_NAME not in await get_supported_features(cql, host)
+
+
+@pytest.mark.asyncio
+async def test_reject_joining_node_without_support_for_enabled_features(manager: ManagerClient) -> None:
```

Comment:
Can we merge this and below test? I don't think there's any value in bootstrapping a completely fresh cluster - it will only prolong CI runs etc. 

We can try the bootstrap, see it fail, and then try the replace in a single test.